doc/decisions/DR024-Versioning-traits-and-specifications-method.md

```diff
@@ -0,0 +1,412 @@
+  mediacreation_version = 3
+else:
+  incompatible_with_manager_error()
+  return
+
+if 1 in supported_versions[companycorp_traits.packageId()]:
+  companycorp_version = 1
+else:
+  incompatible_with_manager_error()
+  return
+
+# Tell the manager up front what versions you are going to use.
+# These must be versions reported from `supportedSchemaVersions`.
+negotiated_schema_versions =
+{
+  openassetio_mediacreation.packageId():mediacreation_version,
+  companycorp_traits.packageId():companycorp_version
+}
+manager.informSchemaVersions(negotiated_schema_versions)
+
+... # Go on to use openassetio methods with trait views of the now
+    # agreed upon version for each trait package.
+```
+### Manager side
+```python
+def supportedSchemaVersions(this, hostSession):
+  # As a manager, we can likely directly declare which schema versions
+  # we support
+  return {
+   openassetio_mediacreation.packageId(), [1,2,3]
+   companycorp_traits.packageId(), [1]
+  }
+
+def informSchemaVersions(this, schemaVersions, hostSession):
+  # Record the schema versions that the host is going to use.
+  # We can use these to branch our internal functionality
+  this.__mediacreation_schema_version = schemaVersions[openassetio_mediacreation.packageId()]
+  this.__companycorp_schema_version = schemaVersions[companycorp_traits.packageId()]
+```
+
+### Pros
+- Decides on version up front, reducing low level trait by trait branching.
+- Versioning is technically optional, which may allow "best-effort" data
+  access. E.g. if the trait ID remains unchanged, and only new
+  properties are added, then a trait version bump is a non-breaking
+  change - `isImbued()` still returns `true` and older trait properties
+  are still accessible.
+- Specifications continue to work as they are currently conceived. If
+  both host and manager agree on the schema version, then the valid
+  selection of Specification view classes is trivially known.
+
+### Cons
+- Complex dance needs to be done up front, people may end up just
+  skipping it.
+- The potential to deliberately skip version negotiation to allow "best
+  effort" access patterns is very bug-prone, may be difficult to debug,
+  and in general introduces confusion about whether versions should be
+  accounted for.
+- Manager reporting a supported schema version could imply every trait
+  in that schema is supported, despite the need to still call
+  managementPolicy.
+- Managers and hosts now have additional data related to specific
+  entities that they must communicate, increasing the complexity of
+  "pass-through" hosts and managers, and complexity of data transfer
+  across multiple hops and/or storage in multiple systems. This could be
+  especially problematic for naive managers that simply wish to be data
+  recorders, as they are almost forced to be version aware.
+- Requires manager plugins to be stateful, at least at `initialze` time.
+  See [no internal state](#no-internal-state).
```

Comment:
[fix] Add:
- Hosts or Managers may not know the versions of packages used in the case of code that works at the TraitsData level (eg BAL, and/or no packages at all), or proxying input from another source. Eg user input.
- Hosts may need/wish to query multiple versions of a particular trait (eg a plugin running inside a DCC with de-coupled code bases).
- Precludes host-host data exchange where the hosts are the only ones that understand the schemas.
- Precludes per-trait versioning which requires prior-knowledge of what traits are compatible between package versions.

Reply:
[1] I don't believe this is a seperate con to saying that there's now additional data that must be communicated and stored, which is a con, but already listed. In the case of non-versioned data from another source, this would hit the same problem even with an in-data approach, a version would have to be invented.
[2] Seems valid
[3] This has never really come up before and i'm not sure it's a legit con. It dosen't preclude this concept, just means you have to write a somewhat generic negotiation mechanism in your manager, which I concede would be harder, but it seems awful niche for a con.
[4] I'm reading this as "Upgrading traits requires out-of-code knowledge of what's compatible", but I find the snippet hard to parse. Seems more like a critique of schema version vs trait by trait versioning, although I guess valid as trait by trait versioning would be harder with an out-of-data formulation. We'd try to solve this with mediacreation utils though.

Reply:
https://github.com/OpenAssetIO/OpenAssetIO/pull/1267/commits/ac7d5b5a95c20e509edbfb2d9330b75b611bd2a7

Reply:
> [1] I don't believe this is a seperate con to saying that there's now additional data that must be communicated and stored, which is a con, but already listed. 

Sorry yeah - this would be a con specific to an "initialization time only" approach as it was attempting to highlight that that information is not always available at initialization time.

> [3] This has never really come up before and i'm not sure it's a legit con. It dosen't preclude this concept, just means you have to write a somewhat generic negotiation mechanism in your manager, which I concede would be harder, but it seems awful niche for a con.

I don't think its that niche, Host A communicating to Host B via a manager as a hub, with proprietary traits is a highly valuable and explicitly supported workflow in the original API design.

Eg. Katana publishes renders with assorted additional "Foundry specific" traits. Nuke then loads those, and uses the Foundry traits to do comp stuff automatically, all with the manager not understanding any of it, other than storing/recalling the traits.  There may well be different versions of those traits published over time for different entities, or the different hosts may bump versions at different times.

How would you see that working with the manager having to negotiate a version in a non-entity specific way?

> [4] I'm reading this as "Upgrading traits requires out-of-code knowledge of what's compatible", but I find the snippet hard to parse. Seems more like a critique of schema version vs trait by trait versioning, although I guess valid as trait by trait versioning would be harder with an out-of-data formulation. We'd try to solve this with mediacreation utils though.

Yeah - it is really scheme vs trait versioning - were you imaging a view of the out-of-data approach where each specific trait is negotiated? I may have missed it but I recall it reading like the out-of-band approach was leaning towards schema versioning - I may well be wrong though.

Reply:
No you wern't misreading, there is a bias in that direction for the out-of-band approach.